File: android/src/main/java/com/nitroplay/video/hybrids/videoviewviewmanager/HybridNitroPlayerViewManager.kt
package com.margelo.nitro.video

import androidx.annotation.OptIn
import androidx.media3.common.util.UnstableApi
import com.facebook.proguard.annotations.DoNotStrip
import com.nitroplay.video.core.NitroPlayerManager
import com.nitroplay.video.core.NitroPlayerViewError
import com.nitroplay.video.core.utils.Threading
import com.nitroplay.video.core.ListenerRegistry

@DoNotStrip
@OptIn(UnstableApi::class)
class HybridNitroPlayerViewManager(nitroId: Int): HybridNitroPlayerViewManagerSpec(), NitroPlayerViewEventsEmitter {
  private var videoView =
    NitroPlayerManager.getNitroPlayerViewWeakReferenceByNitroId(nitroId) ?: throw NitroPlayerViewError.ViewNotFound(nitroId)
  private val registry = ListenerRegistry()
  private var playerDefaults: NitroPlayerDefaults? = null

  init {
    videoView.get()?.eventsEmitter = this
  }

  private fun applyDefaults(player: HybridNitroPlayer?) {
    val defaults = playerDefaults ?: return
    val activePlayer = player ?: return

    defaults.loop?.let { activePlayer.loop = it }
    defaults.muted?.let { activePlayer.muted = it }
    defaults.volume?.let { activePlayer.volume = it }
    defaults.rate?.let { activePlayer.rate = it }
    defaults.mixAudioMode?.let { activePlayer.mixAudioMode = it }
    defaults.ignoreSilentSwitchMode?.let { activePlayer.ignoreSilentSwitchMode = it }
    defaults.playInBackground?.let { activePlayer.playInBackground = it }
    defaults.playWhenInactive?.let { activePlayer.playWhenInactive = it }
  }

  override var player: HybridNitroPlayerSpec?
    get() {
      return Threading.runOnMainThreadSync { return@runOnMainThreadSync videoView.get()?.hybridPlayer }
    }
    set(value) {
      Threading.runOnMainThread {
        val hybridPlayer = value as? HybridNitroPlayer
        videoView.get()?.hybridPlayer = hybridPlayer
        applyDefaults(hybridPlayer)
      }
    }

  override var isAttached: Boolean
    get() = Threading.runOnMainThreadSync { videoView.get()?.isAttachedToWindow == true }
    set(_) {}

  override fun enterFullscreen() {
    videoView.get()?.enterFullscreen()
  }

  override fun exitFullscreen() {
    videoView.get()?.exitFullscreen()
  }

  override var controls: Boolean
    get() = videoView.get()?.useController == true
    set(value) {
      videoView.get()?.useController = value
    }

  override var resizeMode: ResizeMode
    get() = videoView.get()?.resizeMode ?: ResizeMode.NONE
    set(value) {
      videoView.get()?.resizeMode = value
    }

  override var keepScreenAwake: Boolean
    get() = videoView.get()?.keepScreenAwake == true
    set(value) {
      videoView.get()?.keepScreenAwake = value
    }

  override var surfaceType: SurfaceType
    get() = videoView.get()?.surfaceType ?: SurfaceType.SURFACE
    set(value) {
      videoView.get()?.surfaceType = value
    }

  override fun setPlayerDefaults(defaults: NitroPlayerDefaults) {
    playerDefaults = defaults
    applyDefaults(videoView.get()?.hybridPlayer)
  }

  override fun clearPlayerDefaults() {
    playerDefaults = null
  }

  // MARK: - Listener registration methods

  override fun addOnFullscreenChangeListener(listener: (Boolean) -> Unit) =
    registry.add("onFullscreenChange", listener)

  override fun addOnAttachedListener(listener: () -> Unit) =
    registry.add("onAttached", listener)

  override fun addOnDetachedListener(listener: () -> Unit) =
    registry.add("onDetached", listener)

  override fun addWillEnterFullscreenListener(listener: () -> Unit) =
    registry.add("willEnterFullscreen", listener)

  override fun addWillExitFullscreenListener(listener: () -> Unit) =
    registry.add("willExitFullscreen", listener)

  override fun clearAllListeners() {
    registry.clearAll()
  }

  // MARK: - Event emission methods (called by NitroPlayerView)

  override fun onFullscreenChange(isActive: Boolean) =
    registry.emit<(Boolean) -> Unit>("onFullscreenChange") { it(isActive) }

  override fun onAttached() =
    registry.emit<() -> Unit>("onAttached") { it() }

  override fun onDetached() =
    registry.emit<() -> Unit>("onDetached") { it() }

  override fun willEnterFullscreen() =
    registry.emit<() -> Unit>("willEnterFullscreen") { it() }

  override fun willExitFullscreen() =
    registry.emit<() -> Unit>("willExitFullscreen") { it() }

  override val memorySize: Long
    get() = 0
}

interface NitroPlayerViewEventsEmitter {
  fun onAttached()
  fun onDetached()
  fun onFullscreenChange(isActive: Boolean)
  fun willEnterFullscreen()
  fun willExitFullscreen()
}

File: ios/hybrids/NitroPlayerViewManager/HybridNitroPlayerViewManager.swift
//
//  HybridNitroPlayerViewManager.swift
//  NitroPlay
//
//  Created by Krzysztof Moch on 23/09/2024.
//

import Foundation
import AVKit
import NitroModules
import OSLog

struct ViewListenerPair {
  let id: UUID
  let eventName: String
  let callback: Any
}

private let vmLogger = Logger(subsystem: "com.nitroplay.video", category: "ViewManager")

class HybridNitroPlayerViewManager: HybridNitroPlayerViewManagerSpec {
  weak var view: NitroPlayerComponentView?
  var listeners: [ViewListenerPair] = []
  private var playerDefaults: NitroPlayerDefaults?

  let DEALOCATED_WARNING = "NitroPlay: NitroPlayerComponentView is no longer available. It is likely that the view was deallocated."

  init(nitroId: Double) throws {
    guard let view = NitroPlayerComponentView.globalViewsMap.object(forKey: NSNumber(value: nitroId)) else {
      throw NitroPlayerViewError.viewNotFound(nitroId: nitroId).error()
    }

    self.view = view
    super.init()
    view.delegate = NitroPlayerViewDelegate(viewManager: self)
  }

  // MARK: - Private helpers

  private func addListener<T>(eventName: String, listener: T) -> ListenerSubscription {
    let id = UUID()
    listeners.append(ViewListenerPair(id: id, eventName: eventName, callback: listener))
    return ListenerSubscription(remove: { [weak self] in
      self?.listeners.removeAll { $0.id == id }
    })
  }

  private func emitEvent<T>(eventName: String, invoke: (T) throws -> Void) {
    for pair in listeners where pair.eventName == eventName {
      if let callback = pair.callback as? T {
        do {
          try invoke(callback)
        } catch {
          vmLogger.error("[NitroPlay] Error calling \(eventName) listener: \(error)")
        }
      } else {
        vmLogger.error("[NitroPlay] Invalid callback type for \(eventName)")
      }
    }
  }

  private func applyDefaults(to player: HybridNitroPlayer?) {
    guard let player, let defaults = playerDefaults else {
      return
    }

    if let loop = defaults.loop {
      player.loop = loop
    }
    if let muted = defaults.muted {
      player.muted = muted
    }
    if let volume = defaults.volume {
      player.volume = volume
    }
    if let rate = defaults.rate {
      player.rate = rate
    }
    if let mixAudioMode = defaults.mixAudioMode {
      player.mixAudioMode = mixAudioMode
    }
    if let ignoreSilentSwitchMode = defaults.ignoreSilentSwitchMode {
      player.ignoreSilentSwitchMode = ignoreSilentSwitchMode
    }
    if let playInBackground = defaults.playInBackground {
      player.playInBackground = playInBackground
    }
    if let playWhenInactive = defaults.playWhenInactive {
      player.playWhenInactive = playWhenInactive
    }
  }

  // MARK: - Properties

  weak var player: (any HybridNitroPlayerSpec)? {
    get {
      guard let view = view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return nil
      }
      return view.player
    }
    set {
      guard let view = view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return
      }
      view.player = newValue
      applyDefaults(to: newValue as? HybridNitroPlayer)
    }
  }

  var isAttached: Bool {
    get {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return false
      }

      return view.isEffectivelyAttached
    }
    set {
      // Read-only semantic in consumer code; setter exists only to satisfy the generated HybridObject protocol.
    }
  }

  var controls: Bool {
    get {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return false
      }

      return view.controls
    }
    set {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return
      }

      view.controls = newValue
    }
  }

  var resizeMode: ResizeMode {
    get {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return .none
      }

      return view.resizeMode
    }
    set {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return
      }

      view.resizeMode = newValue
    }
  }

  var keepScreenAwake: Bool {
    get {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return false
      }

      return view.keepScreenAwake
    }
    set {
      guard let view else {
        vmLogger.warning("\(self.DEALOCATED_WARNING)")
        return
      }

      view.keepScreenAwake = newValue
    }
  }

  // Android only - no-op on iOS
  var surfaceType: SurfaceType = .surface

  func setPlayerDefaults(defaults: NitroPlayerDefaults) throws {
    playerDefaults = defaults
    applyDefaults(to: player as? HybridNitroPlayer)
  }

  func clearPlayerDefaults() throws {
    playerDefaults = nil
  }

  func enterFullscreen() throws {
    guard let view else {
      throw NitroPlayerViewError.viewIsDeallocated.error()
    }

    try view.enterFullscreen()
  }

  func exitFullscreen() throws {
    guard let view else {
      throw NitroPlayerViewError.viewIsDeallocated.error()
    }

    try view.exitFullscreen()
  }

  // MARK: - Listener registration methods

  func addOnFullscreenChangeListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onFullscreenChange", listener: listener)
  }

  func addOnAttachedListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onAttached", listener: listener)
  }

  func addOnDetachedListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onDetached", listener: listener)
  }

  func addWillEnterFullscreenListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "willEnterFullscreen", listener: listener)
  }

  func addWillExitFullscreenListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "willExitFullscreen", listener: listener)
  }

  func clearAllListeners() throws {
    listeners.removeAll()
  }

  // MARK: - Event emission methods

  func onFullscreenChange(_ isActive: Bool) {
    emitEvent(eventName: "onFullscreenChange") { (callback: (Bool) throws -> Void) in try callback(isActive) }
  }

  func onAttached() {
    emitEvent(eventName: "onAttached") { (callback: () throws -> Void) in try callback() }
  }

  func onDetached() {
    emitEvent(eventName: "onDetached") { (callback: () throws -> Void) in try callback() }
  }

  func willEnterFullscreen() {
    emitEvent(eventName: "willEnterFullscreen") { (callback: () throws -> Void) in try callback() }
  }

  func willExitFullscreen() {
    emitEvent(eventName: "willExitFullscreen") { (callback: () throws -> Void) in try callback() }
  }
}
